### PR TITLE
Bug 1821475 - Show research surface when survey message is available

### PR DESCRIPTION
**Pull request by rocketsroger**
### Pull Request checklist

- [x] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/firefox-android/blob/main/docs/changelog.md) or does not need one
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/firefox-android/blob/main/docs/shared/android/accessibility_guide.md) or does not include any user facing features

### After merge
- **Breaking Changes**: If this is a breaking Android Components change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### To download an APK when reviewing a PR (after all CI tasks finished running):
1. Click on `Checks` at the top of the PR page.
2. Click on the `firefoxci-taskcluster` group on the left to expand all tasks.
3. Click on the `build-apk-{fenix,focus,klar}-debug` task you're interested in.
4. Click on `View task in Taskcluster` in the new `DETAILS` section.
5. The APK links should be on the right side of the screen, named for each CPU architecture.

### GitHub Automation
https://bugzilla.mozilla.org/show_bug.cgi?id=1821475

### CODE REVIEW

File: fenix/app/src/main/java/org/mozilla/fenix/experiments/view/ResearchSurfaceSurvey.kt
Comment (jhugman) [inline fenix/app/src/main/java/org/mozilla/fenix/experiments/view/ResearchSurfaceSurvey.kt:100]: Nit: The optional message `title` is not represented here.

**Review by jhugman — COMMENTED** [review]
Well this is exciting!

I've left some comments about things that need to be fixed, and I'd like to see a UI test to match the [`NimbusMessagingNotificationTest`](https://github.com/mozilla-mobile/firefox-android/blob/daecfa9199d2d33331e8718b3f6338d7db578084/fenix/app/src/androidTest/java/org/mozilla/fenix/ui/NimbusMessagingNotificationTest.kt) and [`NimbusMessagingHomescreenTest`](https://github.com/mozilla-mobile/firefox-android/blob/5b74938c67f5da8a57bf924eedb40736ba8d0ea9/fenix/app/src/androidTest/java/org/mozilla/fenix/ui/NimbusMessagingHomescreenTest.kt).

File: fenix/app/src/main/java/org/mozilla/fenix/HomeActivity.kt
Comment (sarah541) [inline fenix/app/src/main/java/org/mozilla/fenix/HomeActivity.kt:1224]: I am little confused here. Why are we calling `updateMessageAsDisplayed` here when it is automatically called in `onMessageDisplayed` below? This will cause the metadata `displayCount` to be incremented by 2. Not that it matters since the `maxDisplayCount` = 1

File: fenix/app/src/main/java/org/mozilla/fenix/HomeActivity.kt
Comment (rocketsroger) [reply fenix/app/src/main/java/org/mozilla/fenix/HomeActivity.kt:1195]: We do in this case.  Since `FenixMessageSurfaceId.SURVEY` can be used for either the homescreen survey or the fullscreen survey.  The id is the one to determine which one we're showing.

Comment (rocketsroger) [reply fenix/app/src/main/java/org/mozilla/fenix/HomeActivity.kt:1195]: I'll double check.  I was worried that we might use the homescreen survey but you're right this won't happen.

**Comment by rocketsroger** [comment]
> 



> Well this is exciting!
> 
> I've left some comments about things that need to be fixed, and I'd like to see a UI test to match the [`NimbusMessagingNotificationTest`](https://github.com/mozilla-mobile/firefox-android/blob/daecfa9199d2d33331e8718b3f6338d7db578084/fenix/app/src/androidTest/java/org/mozilla/fenix/ui/NimbusMessagingNotificationTest.kt) and [`NimbusMessagingHomescreenTest`](https://github.com/mozilla-mobile/firefox-android/blob/5b74938c67f5da8a57bf924eedb40736ba8d0ea9/fenix/app/src/androidTest/java/org/mozilla/fenix/ui/NimbusMessagingHomescreenTest.kt).
> 
> Thanks for jumping on this so quickly!

Thanks, I'll create an issue to track the work of adding new UI tests for the functionalities we introduced in this feature.

File: fenix/app/src/main/java/org/mozilla/fenix/experiments/view/ResearchSurfaceSurvey.kt
Comment (rocketsroger) [reply fenix/app/src/main/java/org/mozilla/fenix/experiments/view/ResearchSurfaceSurvey.kt:100]: Looking at the [experiment](https://experimenter.services.mozilla.com/nimbus/viewpoint-mobile-messaging-experiment/summary) I don't see title is being used in the experiment.

File: fenix/app/src/main/java/org/mozilla/fenix/HomeActivity.kt
Comment (rocketsroger) [reply fenix/app/src/main/java/org/mozilla/fenix/HomeActivity.kt:1196]: I see what you mean.  Since we're not expected to show it again.  I'll still set it just in case something changed and we didn't notice that it's not set.  That said, I'll move it back since this is not needed until we mark the message as displayed.

File: fenix/app/src/main/java/org/mozilla/fenix/HomeActivity.kt
Comment (rocketsroger) [reply fenix/app/src/main/java/org/mozilla/fenix/HomeActivity.kt]: Let me double check.  Thanks for point this out.

Comment (rocketsroger) [reply fenix/app/src/main/java/org/mozilla/fenix/HomeActivity.kt]: Confirmed that this is not an issue.  Thanks for the help @sarah541

**Comment by rocketsroger** [comment]
> Thanks, I'll create an issue to track the work of adding new UI tests for the functionalities we introduced in this feature.

[Issue](https://bugzilla.mozilla.org/show_bug.cgi?id=1826631) created

**Review by jhugman — APPROVED** [review]
I've left a couple of things that need to be done before this is landed, but I don't want to slow you down with another review cycle:

- write a UI test to demonstrate the surface working (or link to an issue with it on).
- add a title to the card. Since we have multiple surfaces that all work the same, this should work the same too.

**Comment by rocketsroger** [comment]
>     * write a UI test to demonstrate the surface working (or link to an issue with it on).

Thanks, [Issue](https://bugzilla.mozilla.org/show_bug.cgi?id=1826631) created.

File: fenix/app/src/main/java/org/mozilla/fenix/experiments/view/ResearchSurfaceSurvey.kt
Comment (rocketsroger) [reply fenix/app/src/main/java/org/mozilla/fenix/experiments/view/ResearchSurfaceSurvey.kt:100]: I see.  currently we've build this to display something that doesn't have a title to it.  I'll create an issue (not blocking the feature) that tracks the work of supporting title in case it's set.  Thanks

Comment (rocketsroger) [reply fenix/app/src/main/java/org/mozilla/fenix/experiments/view/ResearchSurfaceSurvey.kt:100]: As agreed.  I've created an [issue](https://bugzilla.mozilla.org/show_bug.cgi?id=1826637) to track this as well.  Thanks

File: fenix/app/nimbus.fml.yaml
Comment (Amejia481) [inline fenix/app/nimbus.fml.yaml:90]: Should we create a new style, for this surface?  I guess not all surface will want to show only once, I think this will be a specific behaviour for fullscreen message.

Comment (rocketsroger) [reply fenix/app/nimbus.fml.yaml:90]: This was a decision made as a requirement to use this style.  And since we don't to show it more than once, I think 1 here is the correct value even for the homescreen version of the survey.  Thanks

**Comment by sarah541** [comment]
Question : What happens if the `action` (url in this case) comes out to be empty?

**Comment by rocketsroger** [comment]
> Question : What happens if the `action` (url in this case) comes out to be empty?

Good catch.  I didn't think about this case.  If `action` is empty, then I don't think we should create a `actionUri` and process the intent from that.  I'll fix.

File: fenix/app/src/main/java/org/mozilla/fenix/HomeActivity.kt
Comment (sarah541) [inline fenix/app/src/main/java/org/mozilla/fenix/HomeActivity.kt:1204]:
```suggestion
            processIntent(getIntentForMessage(nextMessage))
```

File: fenix/app/src/main/java/org/mozilla/fenix/HomeActivity.kt
Comment (Amejia481) [inline fenix/app/src/main/java/org/mozilla/fenix/HomeActivity.kt:1188]: nit: Maybe just`showFullscreenMessageIfNeeded` just to be generic :)

File: fenix/app/src/main/java/org/mozilla/fenix/HomeActivity.kt
Comment (Amejia481) [inline fenix/app/src/main/java/org/mozilla/fenix/HomeActivity.kt:282]: No sure if this is the expected behaviour, we will check once, if we haven't loaded the messages before this point we my not show the message.

I'm not sure if should reuse what we have for the home screen message, by dispatching a new action for this new surface.

https://github.com/mozilla-mobile/firefox-android/blob/865b1d75f1376d043538882b6001863f7a22f980/fenix/app/src/main/java/org/mozilla/fenix/messaging/MessagingFeature.kt#L17

and observe [messageToShow](https://github.com/mozilla-mobile/firefox-android/blob/865b1d75f1376d043538882b6001863f7a22f980/fenix/app/src/main/java/org/mozilla/fenix/components/appstate/AppState.kt#L70) changes?

That will warranty that the message will show whenever we re enter the Home Screen, but I'm not sure if this is what we want?

File: fenix/app/src/main/java/org/mozilla/fenix/HomeActivity.kt
Comment (titooan) [inline fenix/app/src/main/java/org/mozilla/fenix/HomeActivity.kt:1229]: I think that defining `showFullscreenMessageSurveyIfNeeded` as `suspend` would avoid the need of using `runBlockingIncrement` here, but maybe I'm missing something.
Does it make sense to you?

File: fenix/app/nimbus.fml.yaml
Comment (Amejia481) [reply fenix/app/nimbus.fml.yaml:90]: Thanks for the clarification 👍🏽

File: fenix/app/src/main/java/org/mozilla/fenix/HomeActivity.kt
Comment (rocketsroger) [reply fenix/app/src/main/java/org/mozilla/fenix/HomeActivity.kt:1204]: Nice, I'll update.  Thanks

File: fenix/app/src/main/java/org/mozilla/fenix/HomeActivity.kt
Comment (rocketsroger) [reply fenix/app/src/main/java/org/mozilla/fenix/HomeActivity.kt:282]: That is a good idea but I don't think it's for this case.  Homescreen messages are handled differently because it's part of the session adapter.  In this case we want to navigate to the full screen message directly.

**Comment by rocketsroger** [comment]
> > Question : What happens if the `action` (url in this case) comes out to be empty?
> 
> Good catch. I didn't think about this case. If `action` is empty, then I don't think we should create a `actionUri` and process the intent from that. I'll fix.

From testing looks like Nimbus does protect us from these issues.  I couldn't get a empty `action` pass in and `action` can't be null either.  I won't do any extra checks then.

**Comment by sarah541** [comment]
> > > Question : What happens if the `action` (url in this case) comes out to be empty?
> > 
> > 
> > Good catch. I didn't think about this case. If `action` is empty, then I don't think we should create a `actionUri` and process the intent from that. I'll fix.
> 
> From testing looks like Nimbus does protect us from these issues. I couldn't get a empty `action` pass in and `action` can't be null either. I won't do any extra checks then.

Even if Nimbus doesn't, I found a check in our code as a fallback. Linking : https://github.com/mozilla-mobile/firefox-android/blob/865b1d75f1376d043538882b6001863f7a22f980/android-components/components/service/nimbus/src/main/java/mozilla/components/service/nimbus/messaging/NimbusMessagingStorage.kt#L165

**Review by sarah541 — APPROVED** [review]

File: fenix/app/src/main/java/org/mozilla/fenix/HomeActivity.kt
Comment (rocketsroger) [reply fenix/app/src/main/java/org/mozilla/fenix/HomeActivity.kt]: Good catch.  Thanks I'll update.

File: fenix/app/src/main/java/org/mozilla/fenix/HomeActivity.kt
Comment (Amejia481) [reply fenix/app/src/main/java/org/mozilla/fenix/HomeActivity.kt]: 👍🏽

**Review by Amejia481 — APPROVED** [review]

**Pull request merged**
